### ISSUE TITLE
[analyzer] False positive core.CallAndMessage on unreachable code guarded by constant global condition

### ISSUE THREAD

**Issue by tianxinghe** [llvm/llvm-project#217963]
clang version 24.0.0git 0c8f1121a2bb2095227dab641e66e5ea4c238e87
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /root/tmptest/llvm-project/build/bin
Build config: +assertions

https://godbolt.org/z/Ghqoa6ros

```c++
    #include <stdio.h>
    
    static int a = 1;
    static int b = 0;
    
    void func1() {
        char *c;
        if (a) {
            printf("%s\n", c);
        }
    }
    
    void func2() {
        char *d;
        if (b) {
            printf("%s\n", d); // false positive
        }
    }
    
    int main() {
        func1();
        func2();
        return 0;
    }
```

Both functions produce warnings:

```
    test.c:9:9: warning: 2nd function call argument is an uninitialized value [core.CallAndMessage]
        9 |         printf("%s\n", c);
          |         ^~~~~~~~~~~~~~~~~
    test.c:16:9: warning: 2nd function call argument is an uninitialized value [core.CallAndMessage]
       16 |         printf("%s\n", d);
          |         ^~~~~~~~~~~~~~~~~
    2 warnings generated.
```

The warning in func2 is a false positive.

**Comment by steakhal** [comment]
The argument is that some other TU could have overwritten the global.
The fact that its static doesnt really help, because some function could have escaped its address anyway. Of course this isnt the case here, but we dont have an analysis that would prove that the global address is never taken.

I find it a fair assumption that if something is constant aka never changes then mark it with const.

**Comment by steakhal** [comment]
I have seen similar FP reports in the past, so it could make some sense to back this with some analysis.

**Comment by tianxinghe** [comment]
> I have seen similar FP reports in the past, so it could make some sense to back this with some analysis.

Thank you for the detailed explanation. I observed that the false positive is eliminated when func1 and its calls are removed. This seems unreasonable. If the analyzer takes a conservative stance that any of these variables could be altered, wouldn't it be logically consistent to also report uninitialized-use warnings for all of them?

**Comment by steakhal** [comment]
Right. You hit a rare edge case here. You probably use it in C mode (thus no global ctors run) and you can see `main`. This suggests the analyzer that the address never could have escaped (because nothing runs before main, and it could inline all functions). This is a super niche case and if you change any of these, you would get both FPs - I'm pretty sure.

**Comment by tianxinghe** [comment]
> Right. You hit a rare edge case here. You probably use it in C mode (thus no global ctors run) and you can see `main`. This suggests the analyzer that the address never could have escaped (because nothing runs before main, and it could inline all functions). This is a super niche case and if you change any of these, you would get both FPs - I'm pretty sure.

https://godbolt.org/z/TE17P59G9
On Godbolt, when I changed both to false, the analyzer still reported only one false positive. However, when I ran the latest version of Clang locally, I got no warning. This seems to indicate that the behavior differs depending on the context.
